Tuna-unit tests for merge_db.py

technicalgrp89 opened the pull request:
Following are the unit test cases added , into test_merge_db_functions.py
  test_parse_jobline
  test_parse_text_fdb_name
  test_parse_text_pdb_name
  test_load_master_list
  test_best_solver
  test_target_merge
  test_update_master_list
  test_write_merge_results
  test_merge_text_file
  test_get_sqlite_table
  test_get_sqlite_row
  test_get_sqlite_data
  test_get_file_list
/utils/coverage_files/coverage_percentage.txt is updated with latest coverage results.

technicalgrp89 committed:
unit tests for merge_db.py

technicalgrp89 committed:
code format error fix

technicalgrp89 commented:
@JehandadKhan @cderb @alexandraBara , Need a code review, Let meknow.

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+        miopenConvolutionFwdAlgoGEMM:GemmFwdRest,0.05712,2749200,miopenConvolutionFwdAlgoGEMM,not used"""
+
+  key, vals = parse_jobline(data)
+  assert key, vals
```
I think a better assertion would be to check for actual expected value like:
assert key = "someExpectedKey" and same for vals

technicalgrp89 replied:
assert are checked for key and vals

cderb replied:
in this case:
assert key == fdb_key

for algorithm in data:
    assert algorithm in vals

for alg_data in data:
    assert alg_data[algorithm] == vals[algorithm]

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+def test_load_master_list():
+  master_file = '/data/tests/gfx803_36.HIP.fdb.txt'
+  master_list = load_master_list(master_file)
+  if master_list == {}:
```
there should be some values in master_list correct? It should check for something other than empty or not

technicalgrp89 replied:
added checks for the content of master list.

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+          'ConvMlirIgemmFwdXdlops,0.03776,0,miopenConvolutionFwdAlgoImplicitGEMM,not used'
+  })
+
+  new_solver, new_time = best_solver(test_value)
```
i think this test would be better if there were 2 solvers in the test_value entries and then the assertion would be checking for the fastest one.
Same for the rest of the best_solver calls bellow

technicalgrp89 replied:
@alexandraBara , I have modified the test more than 1 sampler. where as other tests are checking the bestSolver input values for zero/none for the time value.

technicalgrp89 replied:
best_solver(vals) - vals is set of fdb_key = alg:solver, time, workspace_size, alg; <repeat for algorithms>;...
best_solver -> finds the fastest solver in the passed (input value->Set).Comparing/finding the fastetst solver among more than one Set is outof scope of this functionality. This function it self traverse through the set and returns the appropriate solver/time taken etc.

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+          'ConvMlirIgemmFwdXdlops,0.03776,0,miopenConvolutionFwdAlgoImplicitGEMM,not used'
+  }
+  keep_keys = False
+  target_merge(master_list, key, vals, keep_keys)
```
does this function return anything? there should be an assert after the target_merge() call

cderb replied:
I'd like to have the master list resulting from this merge tested for the new key entry.
If keep_keys is false, I would expect that this would now be the only entry for this key.
There is an opportunity to test keep_keys=True as well. In which case, just this algorithm would be different.

technicalgrp89 replied:
for UT testing purpose, I am supplying only one set of values. We can assert on the basis of result values only.

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+  local_paths = ['/data/tests/../utils/test_files/usr_gfx90a68.HIP.fdb.txt']
+  mids = [-1]
+  keep_keys = False
+  update_master_list(master_list, local_paths, mids, keep_keys)
```
same here, there needs to be a check for return value and some checks that the final data was actually updated correctly. That the best value was chosen etc

technicalgrp89 replied:
assert added to check the result outcome of masterfile

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+  final_file = '/data/tests/old_gfx90a68.HIP.fdb.txt'
+  copy_files = []
+
+  write_merge_results(master_list, final_file, copy_files)
```
this is also missing an assert. We need to check for expected values

technicalgrp89 replied:
Added the the assert to cehck the final file result file content.

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+    assert (False)
+
+  #targetr_file set to None
+  result_file = merge_text_file(master_file=None,
```
missing asserts

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+  cnx_from = sqlite3.connect(local_path)
+  perf_rows, perf_cols = get_sqlite_table(cnx_from, 'perf_db')
+
+  if not perf_rows:
+    assert (False)
+  else:
+    assert (True)
+
+  if not perf_cols:
+    assert (False)
+  else:
+    assert (True)
```
can probably merge this with:
if not perf_rows and perf_cols:
assert (False)
  else:
    assert (True)

technicalgrp89 replied:
corrected

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+    perf = dict(zip(perf_cols, row))
+
+    cfg_row, cfg_cols = get_sqlite_row(cnx_from, 'config', perf['config'])
+    print(type(cfg_row), type(cfg_cols))
```
not printouts pls

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+
+  res, col = get_sqlite_data(cnx_to, 'config', prune_cfg_dims(cfg))
+
+  print(type(res), res)
```
not printouts anywhere pls

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+
+
+def test_get_file_list():
+  print('test - get file list')
```
this...seems to be missing something?

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+this_path = os.path.dirname(__file__)
+
+
+def test_merge_db_functions():
+  test_parse_jobline
+  test_parse_text_fdb_name
+  test_parse_text_pdb_name
+  test_load_master_list
+  test_best_solver
+  test_target_merge
+  test_update_master_list
+  test_write_merge_results
+  test_merge_text_file
+  test_get_sqlite_table
+  test_get_sqlite_row
+  test_get_sqlite_data
+  test_get_file_list
```
this is not necessary, pytest will call every function in this script that starts with 'test_<>'. This can be removed

technicalgrp89 replied:
ok. removed.

alexandraBara commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+import os
+import sqlite3
+
+from tuna.utils.merge_db import parse_jobline
+from tuna.utils.merge_db import parse_text_fdb_name
+from tuna.utils.merge_db import parse_text_pdb_name
+from tuna.utils.merge_db import load_master_list
+from tuna.utils.merge_db import best_solver
+from tuna.utils.merge_db import target_merge
+from tuna.utils.merge_db import no_job_merge
+from tuna.utils.merge_db import single_job_merge
+from tuna.utils.merge_db import multi_job_merge
+from tuna.utils.merge_db import update_master_list
+from tuna.utils.merge_db import write_merge_results
+from tuna.utils.merge_db import merge_text_file
+from tuna.utils.merge_db import merge_sqlite_pdb
+from tuna.utils.merge_db import merge_sqlite_bin_cache
+from tuna.utils.merge_db import merge_sqlite
+from tuna.utils.merge_db import get_file_list
+from tuna.utils.merge_db import get_sqlite_table
+from tuna.utils.merge_db import get_sqlite_row
+from tuna.helper import prune_cfg_dims
+from tuna.utils.merge_db import get_sqlite_data
```
please condense these

technicalgrp89 committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/MITunaX into sp/merge_db_ut

technicalgrp89 committed:
code coverage:11/10/22

alexandraBara reviewed (changes requested):
I dont see the new utest file being called in the groovy file

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+  if (float(new_time)):
+    assert (True)
+  else:
+    assert (False)
```
don't need prints.
the solver string and time should be known based on the test inputs.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+    assert (False)
+
+  #Sample as space, time 0.0
+  test_value = ({
+      'miopenConvolutionFwdAlgoImplicitGEMM':
+          ' ,0.00000,0,miopenConvolutionFwdAlgoImplicitGEMM,not used'
+  })
+
+  new_solver, new_time = best_solver(test_value)
+  if (not (new_solver and new_solver.isspace())):
+    assert (True)
+  else:
+    assert (False)
```
Is this a test for an improperly formatted fdb entry? Why the solver name in particular?

technicalgrp89 replied:
In the above sample, trying to pass the space as the input == solver name.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+
+  #Sample as None and time 0.0
+
+  test_value = ({
+      'miopenConvolutionFwdAlgoImplicitGEMM':
+          'None ,0.00000,0,miopenConvolutionFwdAlgoImplicitGEMM,not used'
+  })
```
Same question, why only testing for improper solver entry?

technicalgrp89 replied:
trying to pass slover name as None

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+                                target_file=None)
+
+
+def test_get_sqlite_table():
+
+  local_path = '/data/utils/test_files/test_gfx90678.db'
+  cnx_from = sqlite3.connect(local_path)
+  perf_rows, perf_cols = get_sqlite_table(cnx_from, 'perf_db')
```
This function is imported from analyze_parse_db.

technicalgrp89 replied:
Yes, You are correct. As it is reffered in merge_db.py, to cover all API , added the test case.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+    assert (True)
+
+
+def test_get_sqlite_row():
+
+  local_path = '/data/utils/test_files/test_gfx90678.db'
+  cnx_from = sqlite3.connect(local_path)
+
+  perf_rows, perf_cols = get_sqlite_table(cnx_from, 'perf_db')
+  for row in perf_rows:
+    perf = dict(zip(perf_cols, row))
+
+    cfg_row, cfg_cols = get_sqlite_row(cnx_from, 'config', perf['config'])
```
function imported from analyze_parse_db

technicalgrp89 replied:
Yes, You are correct. As it is reffered in merge_db.py, to cover all API , added the test case.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,337 @@
+  cnx_from = sqlite3.connect(local_path)
+  cnx_to = sqlite3.connect(final_file)
+
+  res, col = get_sqlite_data(cnx_to, 'config', prune_cfg_dims(cfg))
```
function imported from analyze_parse_db

technicalgrp89 committed:
pytestSuite1()- updated with new merge_db_function tests

technicalgrp89 committed:
PR - 815
	test cases are updated for review comments.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+
+
+def test_parse_text_fdb_name():
+  master_file = '/data/tests/gfx90a_68.HIP.fdb.txt'
+
+  arch, num_cu, final_file, copy_files = parse_text_fdb_name(master_file)
+  assert (arch == "gfx90a")
+  assert (num_cu == '68')
+  assert (final_file == "/data/tests/gfx90a_68.HIP.fdb.txt")
+  assert (copy_files == [])
```
Interesting point to be made here.
Our dbs will read ex gfx900_64, splitting arch and num_cu with a '_' up until 64 cus.
Once we go over that number they get mashed into the arch with hex eg, gfx908 120cus = gfx90878
btw gfx90a68 = gfx90a 104cus
parsing here reacted as expected though.
Perhaps we could test both of those cases in this test.

technicalgrp89 replied:
added gfx90878 db to test the same

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+
+
+def test_parse_text_pdb_name():
+  master_file = '/data/test/gfx1030_36.cd.pdb.txt'
```
This doesn't look like a typical pdb filename 'gfx1030_36.db' is typical.

technicalgrp89 replied:
added new test case to check the pdb num_cu check.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+  })
+
+  solver, time = best_solver(FdbKey_Set)
+  if (len(solver) == 0):
```
could be: assert len(solver)

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+  else:
+    assert (True)
+
+  if (float(time)):
```
could be: assert float(time)
Is this testing if time != 0 ? That appears to be the effect here.

technicalgrp89 replied:
yes, as it is set to 0 in test input

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+
+def test_best_solver():
+
+  FdbKey_Set = ({
```
I don't think this name would pass pylint.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+  })
+
+  new_solver, new_time = best_solver(test_value)
+  if (not (new_solver and new_solver.isspace())):
```
This condition seems overcomplicated.
If you're expecting that new_solver is a space then: assert new_solver.isspace()
If you're expecting that new_solver is None then: assert not new_solver

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+  })
+
+  new_solver, new_time = best_solver(test_value)
+  if (not (new_solver and new_solver == "None")):
```
shorten this to an assert

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,477 @@
+  target_merge(master_list, key, vals, keep_keys)
+
+  #new key with the avaliable values.
+  if (key == '1-160-698-5x5-64-79-348-1-1x1-2x2-1x1-0-NCHW-FP32-F'):
```
Why assert this? This value is set by the test.

technicalgrp89 replied:
updated, by removing same.

technicalgrp89 committed:
- 815 PR - merge_db.py - unit test implementations
   - Review comments incorporated
   - file naming , num_cu checks , assert statements added
   - new unit test cases are added

technicalgrp89 commented:
@cerb review comments updated. Let me know. @alexandraBara

Hildy commented:
> @cerb review comments updated. Let me know. @alexandraBara

I think you meant to @cderb , not @cerb.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+      'miopenConvolutionFwdAlgoImplicitGEMM':
+          'ConvMlirIgemmFwdXdlops,0.03776,0,miopenConvolutionFwdAlgo    ImplicitGEMM,not used'
+  }):
+    assert (True)
```
Don't do this, this will not get the right result.
```py
if (statement) { assert (True) }
```
 
Do this:
```py
assert statement == True
```

technicalgrp89 replied:
statement is populated from master file for that perticular key. Just checking the statement would mislead. So, I would store in temp variable and compare.

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+          'ConvMlirIgemmFwdXdlops,0.03776,0,m    iopenConvolutionFwdAlgoImplicitGEMM,not used'
+  }):
+
+    assert (True)
```
and here

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+  try:
+    target_merge(master_list={}, key={}, vals={}, keep_keys=False)
+  except AttributeError:
+    assert (True)
```
If you are expecting target_merge to end by raising an AttributeError, you should put an 'assert False' after it, if it doesn't fail, the assert will cause a failure.

as a side-note 'assert True' is not useful, it does nothing

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+    target_merge(master_list, key, vals, keep_keys)
+
+  except AttributeError:
+    assert (True)
```
and here

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+
+  if key == open('/data/tests/final.txt').read():
+    assert (True)
+    f.close()
```
What does this test?
If the key is read from the file:
```py
  assert (True) # a null opp
  f.close()     # where is f defined?

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+
+  if key == open('/data/tests/final.txt').read():
+    assert (True)
+    f.close()
```
and here

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+
+  result_file = merge_text_file(master_file, copy_only, keep_keys, target_file)
+
+  if (os.stat(result_file)):
+    assert (True)
+  else:
+    assert (False)
```
simplify to
```py
assert os.stat(result_file)
```

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+  else:
+    assert (False)
+
+  try:
+    assert [row for row in open(master_file)
+           ] == [row for row in open(target_file)]
+  except AssertionError as msg:
+    assert (True)
```
if master == target then the assert is true and this still passes
correct / simplify to
```py
assert statement == False
```

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+
+  master_file = '/data/utils/test_files/old_gfx90a68.HIP.fdb.txt'
+
+  try:
+    result_file = merge_text_file(master_file,
+                                  copy_only=False,
+                                  keep_keys=False,
+                                  target_file=None)
+  except TypeError:
+    assert (True)
```
and here

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+  cnx_from = sqlite3.connect(local_path)
+  perf_rows, perf_cols = get_sqlite_table(cnx_from, 'perf_db')
+
+  if not perf_rows and not perf_cols:
+    assert (False)
+  else:
+    assert (True)
```
simplify to one assert

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+    cfg = dict(zip(cfg_cols, cfg_row))
+    cfg.pop('id', None)
+
+  if not cfg_row and cfg_row:
+    assert (False)
+  else:
+    assert (True)
```
and here

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,481 @@
+
+  res, col = get_sqlite_data(cnx_to, 'config', prune_cfg_dims(cfg))
+
+  if not res:
+    assert (False)
+  else:
+    assert (True)
```
and here

technicalgrp89 committed:
- utils.groovy
   - assert statements are added
   - file operations are added
   - test file path correction

JehandadKhan commented:
@technicalgrp89  Can we please drive this PR to conclusion ?

technicalgrp89 committed:
- #815

    - test file sample .db is added to utils/test_files

    - get_sqlite_data - test sample DB path update

    - merge_db.py : Error handling added to check master_list,
      - key , values, target_file input types

    - test_merge_db_functions.py , added new unit tests to
      - check the invalid input types for
      - target_merge,update_master_list test functions

    - ValueError exception hadling

technicalgrp89 commented:
@cderb , @JehandadKhan CI checks are failing, as we need to support following checks in the source merge_db.py
function - target_merge - key set to None-> raise the exception.
              - merge_text_file , target_file set to None -> raise the exception
                   - update_master_list ,loca_paths set to None - > raise the exception
 rest of the review comments are addressed. Let mw know.

alexandraBara commented:
> @cderb , @JehandadKhan CI checks are failing, as we need to support following checks in the source merge_db.py function - target_merge - key set to None-> raise the exception. - merge_text_file , target_file set to None -> raise the exception - update_master_list ,loca_paths set to None - > raise the exception rest of the review comments are addressed. Let mw know.

@technicalgrp89 all of these failures can be checked in their respective functions. 
Please address all 3 by raising a ValueError('some text here') where each failing condition happens. You can check the Tuna code to see an example where we raise a ValueError.
In the utest you can then catch the ValueError for the failing cases and this should complete the code coverage.

If you have Q's about this please reach out.

technicalgrp89 commented:
test_merge_db_functions.py and tuna/utils/merge_db.py are updated for error handling cases. key, master_list, vals and target_file input values are checked and handled for for Null,None,{}."","   " etc.
@cderb @alexandraBara @JehandadKhan

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,595 @@
+  try:
+    target_merge(master_list, key, vals, keep_keys)
+  except AttributeError as e:
+    assert False
```
Why assert false on attribute error? If the desired outcome is to error on AttributeError, why not just not catch the AttributeError?

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,595 @@
+    assert (file_content.find('miopenConvolutionFwdAlgoImplicitGEMM'))
+
+
+def parse_fdb(filename):
+  find_db = {}
+  fp = open(filename)
+  for line in fp:
+    key, vals = line.split('=')
+    vals = [x.strip() for x in vals.split(';')]
+    find_db[key] = vals
+  fp.close()
+
+  return find_db
```
Maybe this belongs in a utility file?

technicalgrp89 replied:
this is a helper function , used in checking the end result of merge_text_file. Originally, defined in test_merge_db.py

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,595 @@
+  perf_rows, perf_cols = get_sqlite_table(cnx_from, 'perf_db')
+
+  if not perf_rows and not perf_cols:
+    assert (False)
```
equivalent to 
```python
assert perf_rows or perf_cols
```

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,595 @@
+    cfg.pop('id', None)
+
+  if not cfg_row and cfg_row:
+    assert (False)
```
can simplify

cderb commented on tests/test_merge_db_functions.py:
```diff
@@ -0,0 +1,595 @@
+  res, col = get_sqlite_data(cnx_to, 'config', prune_cfg_dims(cfg))
+
+  if not res:
+    assert (False)
```
can simplify

cderb commented on tuna/utils/merge_db.py:
```diff
@@ -207,6 +207,16 @@ def best_solver(vals):
 def target_merge(master_list, key, vals, keep_keys):
   """merge for explicit target file"""
+  #Error handling of Ivalid input parameter types
+  if key is None:
+    raise ValueError(f'Invalid key value: {key}')
+  if (not (key and not key.isspace())):
+    raise ValueError(f'Invalid key value: {key}')
+  if (not (key and key.strip())):
+    raise ValueError(f'Invalid key value: {key}')
```
could simplify to 
```python
if not (key and isinstance(key, str) and key.strip()):
  raise ValueError(f'Invalid key value: {key}')
```
This just asserts that key is a non-empty string. A fairly minimal check.

cderb commented on tuna/utils/merge_db.py:
```diff
@@ -207,6 +207,16 @@ def best_solver(vals):
+    raise ValueError(f'Invalid key value: {key}')
+  if (not (key and key.strip())):
+    raise ValueError(f'Invalid key value: {key}')
+  if not vals:
+    raise ValueError(f'Invalid vals: {vals}')
```
This check will throw an error when vals is None, False, or an empty set. Not very robust, not sure if useful.

cderb commented on tuna/utils/merge_db.py:
```diff
@@ -321,6 +331,12 @@ def write_merge_results(master_list, final_file, copy_files):
+
+  if target_file is None:
+    raise ValueError(f'Invalid key value: {target_file}')
+  if not target_file and target_file.strip():
+    raise ValueError(f'Invalid key value: {target_file}')
```
This conditional is problematic.
if not target_file, then target_file is None, False, 0, "", or empty set
if not target_file, then target_file.strip() will be evaluated, which will cause a separate kind of error on every possible value other than "" on which it will have no affect.
Perhaps the objective was to assert that the target_file is a non empty string or perhaps an existing file.

technicalgrp89 replied:
checking target_file , already exist is not part of UT. test targets for valid i/p type
